model/task_queue_service.go

```diff
@@ -319,54 +333,71 @@ func (t *basicCachedDispatcherImpl) FindNextTask(spec TaskSpec) *TaskQueueItem {
+					"message":                       "schedulableUnit.maxHosts == 0 - this is not a task group; schedulableUnit.tasks is empty - returning nil",
+					"distro_id":                     t.distroID,
+					"schedulableunit_id":            unit.id,
+					"schedulableunit_group":         unit.group,
+					"schedulableunit_project":       unit.project,
+					"schedulableunit_version":       unit.version,
+					"schedulableunit_variant":       unit.variant,
+					"schedulableunit_running_hosts": unit.runningHosts,
+					"schedulableunit_max_hosts":     unit.maxHosts,
+					"schedulableunit_num_tasks":     len(unit.tasks),
+					"task_id_returned":              "",
 				})
 
 				return nil
 			}
 
 			grip.Debug(message.Fields{
-				"ticket":                                "EVG-6289",
-				"function":                              "FindNextTask",
-				"message":                               "schedulableUnit.maxHosts == 0 - this is not a task group",
-				"distro":                                t.distroID,
-				"schedulable_unit_id":                   unit.id,
-				"schedulable_unit_running_hosts":        unit.runningHosts,
-				"schedulable_unit_max_hosts":            unit.maxHosts,
-				"num_schedulable_unit_task_queue_items": len(unit.tasks),
-				"task_id_returned":                      unit.tasks[0].Id,
+				"ticket":                        "EVG-6289",
+				"function":                      "FindNextTask",
+				"message":                       "schedulableUnit.maxHosts == 0 - this is not a task group",
+				"distro_id":                     t.distroID,
+				"schedulableunit_id":            unit.id,
+				"schedulableunit_group":         unit.group,
+				"schedulableunit_project":       unit.project,
+				"schedulableunit_version":       unit.version,
+				"schedulableunit_variant":       unit.variant,
+				"schedulableunit_running_hosts": unit.runningHosts,
+				"schedulableunit_max_hosts":     unit.maxHosts,
+				"schedulableunit_num_tasks":     len(unit.tasks),
+				"task_id_returned":              unit.tasks[0].Id,
 			})
 
 			return &unit.tasks[0]
 		}
+
 		if unit.runningHosts < unit.maxHosts {
 			// TODO: For a multi-host task group, it's not possible to correctly
 			// dispatch tasks based on number of hosts running tasks in the task group
 			// without a transaction. When we use a driver that supports transactions,
 			// we likely want to rewrite this code to use transactions. Currently it's
-			// possible to dispatch a multi-host task group to more tasks than max
+			// possible to dispatch a multi-host task group to more hosts than max
 			// hosts. Before transactions are available, it might be possible to do this
 			// better by performing this query again, just before returning from this
 			// function.
-			numHosts, err = host.NumHostsByTaskSpec(spec.Group, spec.BuildVariant, spec.ProjectID, spec.Version)
+			numHosts, err = host.NumHostsByTaskSpec(unit.group, unit.variant, unit.project, unit.version)
```

Comment:
Should `NumHostsByTaskSpec` return an error if it is passed any zero values?

Reply:
why don't we just pass this a task spec? rather than a lot of arguments?

Reply:
@bsamek - we return an error in this scenario
@tychoish - using `model.TaskSpec{}` here introduces an import cycle